pkg/client/ssa_client.go

```diff
@@ -0,0 +1,218 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/csaupgrade"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ApplyClient the client to use when creating or updating objects. It uses SSA to apply the objects
+// to the cluster and takes care of migrating the objects from ordinary "CRUD" flow to SSA.
+type SsaApplyClient struct {
```

Comment:
I haven't found any fancy "create" function (or anything similar) that could be called without constructing the object "manually".  Something similar to the already existing `NewApplyClient` or `ApplyUnstructuredObjectsWithNewLabels`

Reply:
All the fields need to be set on the struct, so I didn't see the need for the function actually.

Reply:
I like the construct functions more as it's easier to use and it's clear what parameters needs to be set. It could technically also call `GetDefaultFieldOwner` automatically. Or do we expect that we would set something else for this field?

Reply:
I supplied a function to initialize the client from a controller manager, maybe that'll become useful.